File: src/data/articles/examples/q09.ts
import assert from 'node:assert/strict';
import { createHash } from 'node:crypto';
import { mkdtempSync, writeFileSync, readFileSync, rmSync } from 'node:fs';
import { tmpdir } from 'node:os';
import { join } from 'node:path';

function toolOutput(text: string, directory: string, previewChars = 32) {
  const characters = [...text];
  if (characters.length <= previewChars) return { preview: text, artifact: null, truncated: false };
  const name = createHash('sha256').update(text).digest('hex').slice(0, 16) + '.txt';
  // 先写入成功，才能向模型承诺存在可回读的原文。
  writeFileSync(join(directory, name), text, 'utf8');
  return { preview: characters.slice(0, previewChars).join(''), artifact: name, truncated: true };
}

const directory = mkdtempSync(join(tmpdir(), 'context-demo-'));
try {
  const result = toolOutput('connection timeout\n'.repeat(10), directory);
  assert.ok(result.truncated && result.artifact);
  assert.equal(readFileSync(join(directory, result.artifact), 'utf8'), 'connection timeout\n'.repeat(10));
  // 自定义中间表示；适配器仍需转换成供应商的真实图片内容块。
  const blocks = [{ kind: 'text', text: result.preview }, { kind: 'image', source: 'screenshot.png' }];
  assert.equal(blocks[1].kind, 'image');
  console.log(JSON.stringify({ preview_length: [...result.preview].length, kinds: blocks.map(b => b.kind) }));
} finally { rmSync(directory, { recursive: true, force: true }); }

File: src/data/articles/examples/q09.py
import hashlib
import json
import tempfile
from pathlib import Path


def tool_output(text, directory, preview_chars=32):
    if len(text) <= preview_chars:
        return {"preview": text, "artifact": None, "truncated": False}
    name = hashlib.sha256(text.encode()).hexdigest()[:16] + ".txt"
    # 先写入成功，才能向模型承诺存在可回读的原文。
    (Path(directory) / name).write_text(text, encoding="utf-8")
    return {"preview": text[:preview_chars], "artifact": name, "truncated": True}


with tempfile.TemporaryDirectory() as directory:
    result = tool_output("connection timeout\n" * 10, directory)
    assert result["truncated"]
    assert (Path(directory) / result["artifact"]).read_text() == "connection timeout\n" * 10
    # 自定义中间表示；适配器必须把 image 转成供应商的真实图片内容块。
    blocks = [{"kind": "text", "text": result["preview"]}, {"kind": "image", "source": "screenshot.png"}]
    assert blocks[1]["kind"] == "image"
    print(json.dumps({"preview_length": len(result["preview"]), "kinds": [b["kind"] for b in blocks]}))
